actions/install_dpkg_action.go

```diff
@@ -0,0 +1,180 @@
+
+Example to install all packages from recipe subdirectory `pkgs/`:
+
+ - action: install-dpkg
+   description: Install Debian packages from local recipe
+   packages:
+     - pkgs/*.deb
+
+Example to install named packages from recipe subdirectory `pkgs/`:
+
+ - action: install-dpkg
+   description: Install Debian packages from local recipe
+   packages:
+     - pkgs/bmap-tools_*_all.deb
+     - pkgs/fakemachine_*_amd64.deb
+
+Example to download and install a package:
+
+ - action: download
+   description: Install Debian package from url
+   url: http://ftp.us.debian.org/debian/pool/main/b/bmap-tools/bmap-tools_3.5-2_all.deb
+   name: bmap-tools-pkg
+
+ - action: install-dpkg
+   description: Install Debian package from url
+   origin: bmap-tools-pkg
+*/
+
+package actions
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-debos/debos"
+	"github.com/go-debos/debos/wrapper"
+)
+
+type InstallDpkgAction struct {
+	debos.BaseAction `yaml:",inline"`
+	Recommends       bool
+	Unauthenticated  bool
+	Update           bool
+	Origin           string
+	Packages         []string
+}
+
+func NewInstallDpkgAction() *InstallDpkgAction {
+	a := &InstallDpkgAction{Update: true}
+	return a
+}
+
+func (apt *InstallDpkgAction) Run(context *debos.DebosContext) error {
+	aptCommand := wrapper.NewAptCommand(*context, "install-dpkg")
+
+	/* check if named origin exists or fallback to RecipeDir if no origin set */
+	var origin string = context.RecipeDir
+	if len(apt.Origin) > 0 {
+		var found bool
+		if origin, found = context.Origins[apt.Origin]; !found {
+			return fmt.Errorf("origin %s not found", apt.Origin)
+		}
+	}
+
+	/* create a list of full paths of packages to install: if the origin is a
+	 * single file (e.g download action) then just return that package, otherwise
+	 * append package name to the origin path and glob to create a list of packages.
+	 * In other words, install all packages which are in the origin's directory.
+	 */
+	packages := []string{}
+	file, err := os.Stat(origin)
+	if err != nil {
+		return err
+	}
+
+	if file.IsDir() {
+		if len(apt.Packages) == 0 {
+			return fmt.Errorf("no packages defined")
+		}
+
+		for _, pkg := range apt.Packages {
+			// resolve globs
+			source := path.Join(origin, pkg)
+			matches, err := filepath.Glob(source)
+			if err != nil {
+				return err
+			}
+			if len(matches) == 0 {
+				return fmt.Errorf("file(s) not found after globbing: %s", pkg)
+			}
+
+			packages = append(packages, matches...)
+		}
+	} else {
+		packages = append(packages, origin)
+	}
+
+	/* bind mount each package into rootfs & update the list with the
+	 * path relative to the chroot */
+	for idx, pkg := range packages {
+		// check for duplicates after globbing
+		for j := idx + 1; j < len(packages); j++ {
+			if packages[j] == pkg {
+				return fmt.Errorf("duplicate package found: %s", pkg)
+			}
+		}
+
+		log.Printf("Installing %s", pkg)
```

Comment:
you're printing Installing; but there is a bunch of stuff and commands being run before things are actually installed